### ISSUE TITLE
Add "ignore_modules" option to TracebackException.format()

### ISSUE THREAD

**Issue by ncoghlan** [python/cpython#75480]
BPO | [31299](https://bugs.python.org/issue31299)
--- | :---
Nosy | @brettcannon, @ncoghlan, @pitrou, @rbtcollins, @pablogsal, @iritkatriel
PRs | <li>python/cpython#26772</li><li>python/cpython#28067</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-08-29.07:45:45.416>
labels = ['type-feature', 'library', '3.10']
title = 'Add "ignore_modules" option to TracebackException.format()'
updated_at = <Date 2021-08-31.20:42:18.563>
user = 'https://github.com/ncoghlan'
```

bugs.python.org fields:
```python
activity = <Date 2021-08-31.20:42:18.563>
actor = 'iritkatriel'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-08-29.07:45:45.416>
creator = 'ncoghlan'
dependencies = []
files = []
hgrepos = []
issue_num = 31299
keywords = ['patch']
message_count = 10.0
messages = ['300974', '301791', '302892', '304921', '305996', '306762', '328137', '334399', '382694', '400777']
nosy_count = 6.0
nosy_names = ['brett.cannon', 'ncoghlan', 'pitrou', 'rbcollins', 'pablogsal', 'iritkatriel']
pr_nums = ['26772', '28067']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31299'
versions = ['Python 3.10']
```

</p></details>

**Comment by ncoghlan** [comment]
This is a proposed traceback module enhancement deriving from the "python -m" traceback verbosity discussion in bpo-16217.

Specifically, I'm thinking a good way of cleaning up those tracebacks would be to use `TracebackException.format()` to get the formatted exception output, but designate particular modules to be *skipped* during the iteration (specifically `runpy`, `_frozen_importlib`, `importlib`).

I'm filing this as a separate issue so the discussion of the public API for the capability in the traceback module can be separated from the discussion of applying that API in runpy.

**Comment by pitrou** [comment]
I think a more general facility to strip unwanted frames from traceback objects would be useful for more people (including importlib, I think).  i.e. actually mutate the traceback object, not remove the unwanted frames when printing (since it only works if you're the one doing the printing).

(note mutating a traceback object is very easy in C, it's just not possible in pure Python, except perhaps with nasty ctypes hacks)

Also, the `ignore_modules` would be more powerful as a generic `filter` callback, I think.

**Comment by 832af8c7-76e4-4b37-95ba-4c1b2bc490a0** [comment]
Yes, a public built-in facility to modify a traceback would certainly be useful, although there should be a standard way to "hide" unwanted frames without mutating the traceback, too, and the traceback module seems a perfect place to add it.

I have no objections to the `filter` argument, though filtering out frames based on modules to me feels by far more common a task than anything else `filter` would be capable of.

I can update the PR if there's a consensus that `filter` is favored over `ignore_modules`.

**Comment by 832af8c7-76e4-4b37-95ba-4c1b2bc490a0** [comment]
Ping. (this issue needs a decision on ignore_modules vs filter callback, and/or patch review)

**Comment by 832af8c7-76e4-4b37-95ba-4c1b2bc490a0** [comment]
I only recently realized that trace.Trace accepts two similar arguments, namely ignoremods and ignoredirs. Should we try to make the API and implementation of the functionality proposed here conform to that of trace.Trace's ignoremods? Would Python-Ideas be a more appropriate place to discuss this?

**Comment by pablogsal** [comment]
I have prepared a patch (attached) with a possible implementation of the C function that can filter traceback objects modifying the traceback chain. I can make a PR if the patch looks good enough to start iterating over it. In case an alternative implementation is needed I am more than happy to prepare another patch/PR.
